pkg/controller/volume/pvcprotection/pvc_protection_controller.go

```diff
@@ -109,45 +173,64 @@ func (c *Controller) Run(ctx context.Context, workers int) {
 		return
 	}
 
+	go wait.UntilWithContext(ctx, c.runMainWorker, time.Second)
 	for i := 0; i < workers; i++ {
-		go wait.UntilWithContext(ctx, c.runWorker, time.Second)
+		go wait.UntilWithContext(ctx, c.runProcessNamespaceWorker, time.Second)
 	}
 
 	<-ctx.Done()
 }
 
-func (c *Controller) runWorker(ctx context.Context) {
+func (c *Controller) runMainWorker(ctx context.Context) {
 	for c.processNextWorkItem(ctx) {
 	}
 }
 
+func (c *Controller) runProcessNamespaceWorker(ctx context.Context) {
+	for c.processPVCsByNamespace(ctx) {
+	}
+}
+
 // processNextWorkItem deals with one pvcKey off the queue.  It returns false when it's time to quit.
 func (c *Controller) processNextWorkItem(ctx context.Context) bool {
-	pvcKey, quit := c.queue.Get()
-	if quit {
-		return false
+	newNamespaces := make([]string, 0)
+	queueLength := c.queue.Len()
+	for i := 0; i < queueLength; i++ {
+		pvcKey, _ := c.queue.Get()
+		pvcNamespace, pvcName, err := cache.SplitMetaNamespaceKey(pvcKey)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("error parsing PVC key %q: %w", pvcKey, err))
+		}
+		c.pvcProcessingStore.addOrUpdate(pvcNamespace, pvcKey, pvcName, &newNamespaces)
 	}
-	defer c.queue.Done(pvcKey)
-
-	pvcNamespace, pvcName, err := cache.SplitMetaNamespaceKey(pvcKey)
-	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("error parsing PVC key %q: %v", pvcKey, err))
-		return true
+	for _, namespace := range newNamespaces {
+		c.pvcProcessingStore.namespaceQueue.Add(namespace)
```

Comment:
Good catch! This logic was probably left over from when I was keeping the namespace queue and namespaceToPVCsMap separate instead of in shared structure.